pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/reader/ForwardIndexReader.java

```diff
@@ -916,4 +919,67 @@ default byte[][] getBytesMV(int docId, T context) {
   default int getNumValuesMV(int docId, T context) {
     throw new UnsupportedOperationException();
   }
+
+  /**
+   * This class represents the buffer byte ranges accessed while reading a given docId.
+   * See {@link ValueRangeProvider} for more details.
+   */
+  @AllArgsConstructor
+  @EqualsAndHashCode
+  @Getter
+  class ValueRange {
+    private final long _offset;
+    private final int _sizeInBytes;
+
+    public static ValueRange newByteRange(long offset, int sizeInBytes) {
+      return new ValueRange(offset, sizeInBytes);
+    }
+
+    @Override
+    public String toString() {
+      return "Range{" + "_offset=" + _offset + ", _size=" + _sizeInBytes + '}';
+    }
+  }
+
+  /**
+   * An interface that enables the caller to get byte ranges accessed while reading a given docId.
+   * This can be used to eventually allow prefetching of exact byte ranges during query processing,
+   * if the docIds to be fetched are known beforehand.
+   *
+   * This interface will be implemented by all the forward index readers.
+   */
+  interface ValueRangeProvider<T extends ForwardIndexReaderContext> {
+    /**
+     * Returns a list of {@link ValueRange} that represents all the distinct
+     * buffer byte ranges (absolute offset, sizeInBytes) that are accessed when reading the given (@param docId}
+     * @param docId to find the range for
+     * @param context Reader context
+     * @param ranges List of {@link ValueRange} to which the applicable value ranges will be added
+     */
+    void recordDocIdByteRanges(int docId, T context, List<ValueRange> ranges);
```

Comment:
Will the caller use this method when the value is fixed length?

Reply:
This method is not supposed to be used if value is fixed length type. Most implementations throw unsupported exception if using this on a fixed offset type fwd index